File: server/src/db/entities.rs
pub mod api_keys {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "api_keys")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub tenant_id: Option<String>,
        pub name: String,
        pub key_hash: String,
        pub prefix: String,
        pub enabled: i32,
        pub scopes_json: String,
        pub model_allowlist_json: String,
        pub channel_allowlist_json: String,
        pub token_limit_daily: Option<i64>,
        pub rate_limit_per_minute: Option<i64>,
        pub expires_at: Option<String>,
        pub created_at: String,
        pub updated_at: String,
        pub last_used_at: Option<String>,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

#[allow(dead_code)]
pub mod app_settings {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "app_settings")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub key: String,
        pub value: String,
        pub updated_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

pub mod channels {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "channels")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub name: String,
        pub base_url: String,
        pub credential_id: Option<String>,
        pub enabled: i32,
        pub priority: i32,
        pub weight: i32,
        pub model_allowlist_json: String,
        pub status: String,
        pub health_score: i32,
        pub cooldown_until: Option<String>,
        pub last_error: Option<String>,
        pub last_used_at: Option<String>,
        pub created_at: String,
        pub updated_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

pub mod codex_credentials {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "codex_credentials")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub label: String,
        pub email: String,
        pub account_id: String,
        pub plan_type: String,
        pub token_ciphertext: String,
        pub token_expires_at: Option<String>,
        pub enabled: i32,
        pub priority: i32,
        pub weight: i32,
        pub user_agent: Option<String>,
        pub upstream_transport: String,
        pub proxy_json: Option<String>,
        pub metadata_json: String,
        pub last_refresh_at: Option<String>,
        pub last_used_at: Option<String>,
        pub created_at: String,
        pub updated_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

#[allow(dead_code)]
pub mod codex_quota_cache {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "codex_quota_cache")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub credential_id: String,
        pub status: String,
        pub cache_json: String,
        pub retrieved_at: String,
        pub updated_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

pub mod oauth_pending_states {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "oauth_pending_states")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub state: String,
        pub provider: String,
        pub code_verifier: String,
        pub code_challenge: String,
        pub redirect_uri: String,
        pub created_at: String,
        pub expires_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

pub mod request_logs {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "request_logs")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub started_at: String,
        pub completed_at: String,
        pub method: String,
        pub path: String,
        pub request_type: String,
        pub stream: i32,
        pub model: String,
        pub status_code: i32,
        pub latency_ms: i64,
        pub api_key_id: Option<String>,
        pub api_key_prefix: Option<String>,
        pub channel_id: Option<String>,
        pub credential_id: Option<String>,
        pub prompt_tokens: i64,
        pub completion_tokens: i64,
        pub total_tokens: i64,
        pub error_code: Option<String>,
        pub error_message: Option<String>,
        pub request_body_text: Option<String>,
        pub upstream_body_text: Option<String>,
        pub timing_json: Option<String>,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

#[allow(dead_code)]
pub mod tenants {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "tenants")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub name: String,
        pub owner_email: String,
        pub enabled: i32,
        pub max_api_keys: Option<i64>,
        pub token_limit_daily: Option<i64>,
        pub rate_limit_per_minute: Option<i64>,
        pub model_allowlist_json: String,
        pub channel_allowlist_json: String,
        pub allow_custom_proxy: i32,
        pub allow_custom_user_agent: i32,
        pub proxy_json: Option<String>,
        pub user_agent: Option<String>,
        pub expires_at: Option<String>,
        pub metadata_json: String,
        pub created_at: String,
        pub updated_at: String,
        pub deleted_at: Option<String>,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

#[allow(dead_code)]
pub mod channel_credentials {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "channel_credentials")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub channel_id: String,
        #[sea_orm(primary_key, auto_increment = false)]
        pub credential_id: String,
        pub created_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

pub mod proxy_pool {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "proxy_pool")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub name: String,
        pub r#type: String,
        pub host: String,
        pub port: i32,
        pub username: String,
        pub password_ciphertext: Option<String>,
        pub enabled: i32,
        pub notes: String,
        pub created_at: String,
        pub updated_at: String,
        pub last_used_at: Option<String>,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

pub mod tenant_users {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "tenant_users")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub tenant_id: String,
        pub email: String,
        pub display_name: String,
        pub role: String,
        pub enabled: i32,
        pub password_hash: Option<String>,
        pub last_login_at: Option<String>,
        pub created_at: String,
        pub updated_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

pub mod tenant_invites {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "tenant_invites")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub tenant_id: String,
        pub user_id: String,
        pub email: String,
        pub token_hash: String,
        pub expires_at: String,
        pub accepted_at: Option<String>,
        pub revoked_at: Option<String>,
        pub created_at: String,
        pub updated_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

#[allow(dead_code)]
pub mod request_log_details {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "request_log_details")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub request_log_id: String,
        pub created_at: String,
        pub updated_at: String,
        pub request_headers_json: Option<String>,
        pub request_body_text: Option<String>,
        pub forwarded_body_text: Option<String>,
        pub upstream_headers_json: Option<String>,
        pub upstream_body_text: Option<String>,
        pub error_details_json: Option<String>,
        pub stage_timings_json: Option<String>,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

#[allow(dead_code)]
pub mod usage_records {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "usage_records")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub request_log_id: Option<String>,
        pub api_key_id: Option<String>,
        pub api_key_prefix: Option<String>,
        pub api_key_name: Option<String>,
        pub tenant_id: Option<String>,
        pub tenant_name: Option<String>,
        pub channel_id: Option<String>,
        pub credential_id: Option<String>,
        pub model: String,
        pub request_type: String,
        pub prompt_tokens: i64,
        pub completion_tokens: i64,
        pub total_tokens: i64,
        pub cached_tokens: i64,
        pub created_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

#[allow(dead_code)]
pub mod channel_health_events {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "channel_health_events")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub channel_id: String,
        pub credential_id: Option<String>,
        pub status: String,
        pub status_code: Option<i32>,
        pub message: Option<String>,
        pub created_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

#[allow(dead_code)]
pub mod audit_logs {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, Eq, DeriveEntityModel)]
    #[sea_orm(table_name = "audit_logs")]
    pub struct Model {
        #[sea_orm(primary_key)]
        pub id: String,
        pub actor_type: String,
        pub actor_id: Option<String>,
        pub operation: String,
        pub target_type: Option<String>,
        pub target_id: Option<String>,
        pub metadata_json: String,
        pub created_at: String,
    }

    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}

    impl ActiveModelBehavior for ActiveModel {}
}

File: src/shared/types/entities.ts
export type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };

export type ChannelStatus =
  | "healthy"
  | "degraded"
  | "cooling_down"
  | "disabled";

export type CodexAccountUsageStatus = "normal" | "warning" | "error" | "unused";

export interface CodexAccountUsageHealth {
  status: CodexAccountUsageStatus;
  score: number;
  requestCount: number;
  successCount: number;
  errorCount: number;
  lastUsedAt: string | null;
  lastStatusCode: number | null;
  lastErrorCode: string | null;
  windowSize: number;
}

export interface ApiKeyRecord {
  id: string;
  tenantId: string | null;
  name: string;
  prefix: string;
  keyHash: string;
  scopes: string[];
  modelAllowlist: string[];
  channelAllowlist: string[];
  enabled: boolean;
  tokenLimitDaily: number | null;
  rateLimitPerMinute: number | null;
  expiresAt: string | null;
  createdAt: string;
  updatedAt: string;
  lastUsedAt: string | null;
}

export interface PublicApiKey {
  id: string;
  tenantId: string | null;
  name: string;
  prefix: string;
  scopes: string[];
  modelAllowlist: string[];
  channelAllowlist: string[];
  enabled: boolean;
  tokenLimitDaily: number | null;
  rateLimitPerMinute: number | null;
  expiresAt: string | null;
  createdAt: string;
  updatedAt: string;
  lastUsedAt: string | null;
}

export interface CreatedApiKey extends PublicApiKey {
  key: string;
}

export interface TenantLimits {
  quotaShares: number | null;
  maxApiKeys: number | null;
  tokenLimitDaily: number | null;
  rateLimitPerMinute: number | null;
  modelAllowlist: string[];
  channelAllowlist: string[];
  expiresAt: string | null;
}

export interface TenantRecord extends TenantLimits {
  id: string;
  name: string;
  ownerEmail: string;
  enabled: boolean;
  allowCustomProxy: boolean;
  allowCustomUserAgent: boolean;
  proxy: PublicCredentialProxyConfig | null;
  userAgent: string | null;
  createdAt: string;
  updatedAt: string;
  deletedAt: string | null;
  metadata: Record<string, unknown>;
}

export interface TenantWithSecrets
  extends Omit<TenantRecord, "proxy"> {
  proxy: CredentialProxyConfig | null;
}

export interface PublicTenant extends TenantRecord {
  apiKeyCount: number;
  enabledApiKeyCount: number;
  todayTokens: number;
  pendingInvite: boolean;
  lastLoginAt: string | null;
  passwordChangedAt: string | null;
}

export interface TenantUserRecord {
  id: string;
  tenantId: string;
  email: string;
  displayName: string;
  role: "owner";
  enabled: boolean;
  passwordHash: string | null;
  lastLoginAt: string | null;
  passwordChangedAt: string | null;
  sessionVersion: number;
  createdAt: string;
  updatedAt: string;
}

export interface TenantPasswordResetRecord {
  id: string;
  tenantId: string;
  userId: string;
  tokenHash: string;
  expiresAt: string;
  consumedAt: string | null;
  revokedAt: string | null;
  createdAt: string;
}

export interface CreatedTenantPasswordReset {
  token: string;
  resetPath: string;
  expiresAt: string;
}

export interface TenantInviteRecord {
  id: string;
  tenantId: string;
  userId: string | null;
  email: string;
  tokenHash: string;
  expiresAt: string;
  acceptedAt: string | null;
  revokedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CreatedTenantInvite {
  id: string;
  tenantId: string;
  email: string;
  expiresAt: string;
  activateUrl: string;
  token: string;
}

export interface TenantResourceChannel {
  id: string;
  name: string;
  enabled: boolean;
  status: ChannelStatus;
  modelAllowlist: string[];
  credentialIds: string[];
}

export interface TenantResourceCredential {
  id: string;
  provider: "codex";
  email: string;
  accountId: string;
  planType: string;
  enabled: boolean;
  priority: number;
  weight: number;
  fastEnabled: boolean;
  upstreamTransport: CodexUpstreamTransport;
  useGlobalProxy: boolean;
  proxy: PublicCredentialProxyConfig | null;
  usageHealth?: CodexAccountUsageHealth;
  expiresAt: string | null;
  lastRefreshAt: string | null;
  lastUsedAt: string | null;
  cooldownUntil: string | null;
  lastError: string | null;
}

export interface TenantResources {
  models: string[];
  channels: TenantResourceChannel[];
  credentials: TenantResourceCredential[];
}

export interface TenantRuntimeContext {
  id: string;
  name: string;
  proxy: CredentialProxyConfig | null;
  userAgent: string | null;
  quotaShares: number | null;
}

export type CredentialProxyType = "socks5" | "socks5h";

export interface CredentialProxyConfig {
  enabled: boolean;
  type: CredentialProxyType;
  host: string;
  port: number;
  username: string;
  password: string;
}

export interface PublicCredentialProxyConfig {
  enabled: boolean;
  type: CredentialProxyType;
  host: string;
  port: number;
  username: string;
  passwordSet?: boolean;
}

export interface ProxyPoolRecord {
  id: string;
  name: string;
  enabled: boolean;
  type: CredentialProxyType;
  host: string;
  port: number;
  username: string;
  passwordSet: boolean;
  notes: string;
  createdAt: string;
  updatedAt: string;
  lastUsedAt: string | null;
}

export interface ProxyPoolRecordWithSecret extends Omit<
  ProxyPoolRecord,
  "passwordSet"
> {
  password: string;
}

export type CodexUpstreamTransport = "http" | "websocket";

export interface CodexTokenBundle {
  access_token: string;
  refresh_token: string;
  id_token: string;
  expired: string;
  last_refresh: string;
}

export interface CodexCredentialRecord {
  id: string;
  provider: "codex";
  email: string;
  accountId: string;
  planType: string;
  enabled: boolean;
  priority: number;
  weight: number;
  fastEnabled: boolean;
  upstreamTransport: CodexUpstreamTransport;
  userAgent: string | null;
  useGlobalProxy: boolean;
  proxyPoolId: string | null;
  proxy: PublicCredentialProxyConfig | null;
  usageHealth?: CodexAccountUsageHealth;
  expiresAt: string | null;
  lastRefreshAt: string | null;
  lastUsedAt: string | null;
  cooldownUntil: string | null;
  lastError: string | null;
  createdAt: string;
  updatedAt: string;
  metadata: Record<string, unknown>;
}

export type CodexCredentialWithTokens = Omit<CodexCredentialRecord, "proxy"> & {
  proxy: CredentialProxyConfig | null;
  tokens: CodexTokenBundle;
};

export interface GlobalSettingsRecord {
  publicBaseUrl: string;
  proxy: PublicCredentialProxyConfig | null;
  proxySource: "database" | "environment" | "none";
  userAgent: string;
  userAgentSource: "database" | "environment" | "default";
  fullRequestLoggingEnabled: boolean;
  codexAutoDisableRefreshExhausted: boolean;
  requestLogRetentionDays: number | null;
  requestLogDetailRetentionDays: number | null;
  timeZone: string;
  timeZonePending: string | null;
  timeZoneRebuildStatus: TimeZoneRebuildStatus;
  timeZoneRebuildError: string | null;
  updatedAt: string | null;
}

export type TimeZoneRebuildStatus =
  | "idle"
  | "pending"
  | "running"
  | "failed";

export interface ChannelRecord {
  id: string;
  name: string;
  provider: "codex";
  baseUrl: string;
  credentialId: string;
  credentialIds: string[];
  enabled: boolean;
  priority: number;
  weight: number;
  modelAllowlist: string[];
  status: ChannelStatus;
  healthScore: number;
  usageHealth?: CodexAccountUsageHealth;
  cooldownUntil: string | null;
  lastError: string | null;
  lastUsedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface RelayApiKeyContext {
  id: string;
  tenantId: string | null;
  tenant: TenantRuntimeContext | null;
  name: string;
  prefix: string;
  scopes: string[];
  modelAllowlist: string[];
  channelAllowlist: string[];
  tokenLimitDaily: number | null;
  rateLimitPerMinute: number | null;
}

export interface RelayRequestContext {
  apiKey: RelayApiKeyContext;
  model: string;
  requestType: string;
  stream: boolean;
  method: string;
  path: string;
}

export interface SelectedChannel {
  channel: ChannelRecord;
  credential: CodexCredentialWithTokens;
}

export interface UsageSnapshot {
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  cachedTokens: number;
  cacheWriteTokens?: number;
  reasoningTokens?: number;
  costNanoUsd?: string | null;
  priceModel?: string | null;
  priceVersion?: string | null;
  pricingComplete?: boolean;
}

export interface UsageStatsRow {
  key: string;
  label: string;
  subLabel: string | null;
  requestCount: number;
  successCount: number;
  errorCount: number;
  streamCount: number;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  cachedTokens: number;
  cacheHitRate: number;
  avgLatencyMs: number;
  p95LatencyMs: number;
  avgFirstTokenLatencyMs: number;
  p95FirstTokenLatencyMs: number;
  avgTokensPerRequest: number;
  tokensPerSecond: number;
  firstRequestAt: string | null;
  lastRequestAt: string | null;
}

export interface ApiKeyUsageStatsRow extends UsageStatsRow {
  apiKeyId: string | null;
  apiKeyPrefix: string | null;
  apiKeyName: string;
  enabled: boolean | null;
  tokenLimitDaily: number | null;
  todayTokens: number;
  tokenLimitUtilization: number | null;
}

export interface TenantUsageStatsRow extends UsageStatsRow {
  tenantId: string | null;
  tenantName: string;
  enabled: boolean | null;
  tokenLimitDaily: number | null;
  todayTokens: number;
  tokenLimitUtilization: number | null;
}

export interface ApiKeyModelUsageStatsRow extends UsageStatsRow {
  apiKeyId: string | null;
  apiKeyPrefix: string | null;
  apiKeyName: string;
  model: string;
}

export interface ApiKeyDailyUsageStatsRow extends DailyUsageStatsRow {
  apiKeyId: string | null;
  apiKeyPrefix: string | null;
  apiKeyName: string;
}

export interface DailyDimensionUsageStatsRow extends UsageStatsRow {
  date: string;
  dimension: "tenant" | "api_key" | "model" | "channel" | "credential" | "request_type";
  dimensionId: string | null;
  dimensionName: string;
}

export interface ErrorCodeDailyStatsRow {
  date: string;
  errorCode: string;
  requestCount: number;
  tenantId: string | null;
  tenantName: string | null;
}

export interface DailyUsageStatsRow {
  date: string;
  requestCount: number;
  successCount: number;
  errorCount: number;
  streamCount: number;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  cachedTokens: number;
  cacheHitRate: number;
  avgLatencyMs: number;
  p95LatencyMs: number;
  avgFirstTokenLatencyMs: number;
  p95FirstTokenLatencyMs: number;
  avgTokensPerRequest: number;
  tokensPerSecond: number;
}

export interface ActivityHeatmapDay {
  date: string;
  requestCount: number;
  successCount: number;
  errorCount: number;
  streamCount: number;
  totalTokens: number;
  level: number;
}

export interface ActivityHeatmapStats {
  generatedAt: string;
  scope: "site" | "api_key";
  apiKeyId: string | null;
  apiKeyName: string | null;
  apiKeyPrefix: string | null;
  from: string;
  to: string;
  weeks: number;
  days: ActivityHeatmapDay[];
  totalRequests: number;
  totalTokens: number;
  activeDays: number;
  maxRequests: number;
  currentStreakDays: number;
  longestStreakDays: number;
}

export interface AdminOverviewTotals {
  requestCount: number;
  successCount: number;
  errorCount: number;
  streamCount: number;
  promptTokens: number;
  completionTokens: number;
  totalTokens: number;
  cachedTokens: number;
  cacheHitRate: number;
  avgLatencyMs: number;
  p95LatencyMs: number;
  avgFirstTokenLatencyMs: number;
  p95FirstTokenLatencyMs: number;
  avgTokensPerRequest: number;
  tokensPerSecond: number;
  distinctApiKeyCount: number;
  distinctModelCount: number;
  distinctChannelCount: number;
  firstRequestAt: string | null;
  lastRequestAt: string | null;
}

export interface AdminOverviewRange {
  from: string;
  to: string;
  days: number;
}

export interface AdminOverviewAnomaly {
  id: string;
  severity: "info" | "warning" | "critical";
  category: "error" | "latency" | "quota" | "traffic" | "routing";
  title: string;
  description: string;
  date: string | null;
  tenantId: string | null;
  tenantName: string | null;
  targetId: string | null;
  targetName: string | null;
  metric: string;
  value: number;
  baseline: number | null;
}

export interface AdminOverviewStats {
  generatedAt: string;
  range: AdminOverviewRange;
  totals: AdminOverviewTotals;
  byTenant: TenantUsageStatsRow[];
  byApiKey: ApiKeyUsageStatsRow[];
  byApiKeyDay: ApiKeyDailyUsageStatsRow[];
  byApiKeyModel: ApiKeyModelUsageStatsRow[];
  byModel: UsageStatsRow[];
  byChannel: UsageStatsRow[];
  byCredential: UsageStatsRow[];
  byRequestType: UsageStatsRow[];
  byDay: DailyUsageStatsRow[];
  byTenantDay: DailyDimensionUsageStatsRow[];
  byModelDay: DailyDimensionUsageStatsRow[];
  byChannelDay: DailyDimensionUsageStatsRow[];
  byCredentialDay: DailyDimensionUsageStatsRow[];
  byRequestTypeDay: DailyDimensionUsageStatsRow[];
  byErrorCodeDay: ErrorCodeDailyStatsRow[];
  anomalies: AdminOverviewAnomaly[];
}
